usr/src/man/btref.1
'\"macro stdmacro
.nr X
.TH btref 1 "29 Dec 1994"
.SH NAME
btref \- BattleTris Referee management software
.SH SYNOPSIS
.B btref
[
.B \-h
|
.B \-?
]
.SH DESCRIPTION
.IX "btref" "" "\fLbtref\fP \(em btref"
.IX "btref" "BattleTris" \(em \fLbtref\fP"
.BR BattleTris (1)
is a two-player networked game based on
.I Tetris\^.
Player and network information is stored in shared database files.
This database is maintained by separate
.I BattleTris Server\^
software.
The
.BR btref (1)
.I BattleTris Referee Software\^
allows one or more administrators to manage these database files
and browse their contents.
.PD
.SH USING THE REFEREE
In order to use the referee, the administrator must be logged on to
a BattleTris Server host.  This is simply a host on which the database
files are stored locally.  The database files must be stored on a local
disk on the BattleTris Server in order for proper functioning of the
BattleTris server daemons.  Refer to
.BR btserverd (1)
for more information on database file configuration.
.PP
When the referee software is launched, it will attempt to open both
the network and player database files.  If you plan to have more than
one individual who acts as a BattleTris administrator, you should
create a UNIX group for these users, and
.BR chmod (1)
the database files to be readable and writable by members of this group.
If the referee software fails to open any of the needed database files,
it will report an error and exit.
.PP
Once the referee software is initialized, you may enter commands at the
prompt.  A list of commands is presented later in this documentation, and
is available within the referee by giving the
.I help\^
command.  Commands may be abbreviated to the shortest unique command name.
For example, entering
.I nl\^
will result in executing the command
.I nlist\^
.  Also, some of the referee commands support pattern arguments.  Patterns
are given using standard shell ``globbing'' syntax.  Refer to the section on
.I Filename Generation\^
in
.BR sh (1)
if you are unfamiliar with this syntax.
.PP
.I IMPORTANT:\^
You should not execute commands which flush or compress the database files
while the BattleTris server daemons are running.  All commands which
modify the database files will ask you to confirm them before proceeding.
If standard input to the referee is not a terminal (but instead a file or
shell ``here-document'') no prompts will be printed and the referee will
not attempt to confirm any commands.  Finally, make regular backups of
all database files, and always make a backup before executing the compression
commands, as these are volatile operations.
.PD
.SH OPTIONS
.TP 10
.B \-h
Display usage information.
btref will print out a short synopsis of available command-line options.
.PD
.SH COMMANDS
.PD 0
.TP 5
.B nlist
Print a short listing of the users currently using BattleTris, the host
they are playing from, and their current status.
.TP
.B ndata
Print a verbose listing of all the data for entries in the network database.
This includes a database key for each entry.  This key is needed to delete
entries from the network database if an entry is not removed properly by
the server software.
.TP
.B ndelete <key>
Delete the network database entry with the specified key.
.TP
.B nflush
Delete all entries in the network database and truncate the network database
files.
.TP
.B ncruft <threshold>
List the network entries which are older than the specified threshold (in
hours).  The administrator may use this information to determine suspect
cruft entries.  Beware: some users may be able to play BattleTris for large
numbers of hours without mental collapse.  This command does not actually
verify a connection to any BattleTris clients, so use caution before
removing anyone from the network database.
.TP
.B nclean <threshold>
Remove the network entries which are older than the specified threshold (in
hours).  See the warnings regarding ``ncruft'' above.
.TP
.B ncompress
Compress the network database.  Because of the algorithm used for the
database files in BattleTris, the database files may grow quite large on
disk, even though the number of entries remains constant.  This is more
of a problem with the player database, because the entries are constantly
changing size, and so the freelist of records inside the database file
grows quite large, wasting space inside the file.  This command removes
all the unused free space from the network database files.
.TP
.B plist <pattern>
Print a listing of the login names of all users currently in the player
database.
.TP
.B pdata <pattern>
Print a listing of player information for each player whose login name
matches the specified pattern.  The information displayed includes all
of their statistics, and is equivalent to the information displayed for
each player in the BattleTris Challenge or BattleTris Roster screens.
.TP
.B pdelete <pattern>
Delete the player entries for each player whose login name matches the
specified pattern.  This is useful at Brown because Computer Science
students are given transitory accounts for a course which are removed
at the end of the semester.  Since someone else may be reassigned this
account number in the future, we want to get rid of all course accounts
at the end of each semester.
.TP
.B pflush
Delete all entries in the player database and truncate the player database
files.
.TP
.B pcompress
Compress the player database.  Refer to the description of ``ncompress''
above for information about database compression.
.TP
.B stats
Compute the number of players currently stored in the player database and
an approximate number of games played.
.TP
.B help [command1 command2 ...]
Print a summary of all the referee commands and their usage.
.TP
.B quit
Quit the referee.
.PD
.SH FILES
.PD 0
.TP 5
.B $HOME/.btplan
Your own personalized BattleTris plan file.  The first line specifies
your nickname, and the other lines specify your plan, as shown in
the roster screen.  Make sure to
.BR chmod (1)
your BattleTris plan file to be have read permissions for "other".
.TP
.B /cs/src/BattleTris/bin/btref
BattleTris Referee application software.
.TP
.B /cs/app-defaults/BattleTris
BattleTris X-Window default resources.
.TP
.B /cs/man/man1/BattleTris.1
BattleTris documentation.
.TP
.B /cs/man/man1/btref.1
BattleTris referee documentation.
.PD
.SH AUTHORS
Bryan Cantrill, Charles Hoecker, and Michael Shapiro
.TP
Brown University Department of Computer Science


Codemgr_wsdata/history
VERSION 1
COMMAND bringover -p /net/kiowa/export/BattleTris -w bt-ppc usr 
create: usr/bin/install-sh
create: usr/src/BTConfig.H
create: usr/src/BTConfig.H.bot
create: usr/src/BTConfig.H.in
create: usr/src/BTConfig.H.top
create: usr/src/Makefile
create: usr/src/Makeinclude
create: usr/src/Makeinclude.in
create: usr/src/acconfig.h
create: usr/src/config.cache
create: usr/src/config.log
create: usr/src/config.status
create: usr/src/configure
create: usr/src/configure.in
create: usr/src/art/btbazaar.ppm
create: usr/src/art/btbiff1.ppm
create: usr/src/art/btbiff3.ppm
create: usr/src/art/btbiff4.ppm
create: usr/src/art/btchalbiff.ppm
create: usr/src/art/btchalbiff.xpm
create: usr/src/art/btchallbiff.xpm
create: usr/src/art/btchallmask.xbm
create: usr/src/art/btchalmask.xbm
create: usr/src/art/btchalmask.xbm.old
create: usr/src/art/btchalmask.xpm
create: usr/src/art/btchalmask2.xbm
create: usr/src/art/btgimp.ppm
create: usr/src/art/btgimp2.ppm
create: usr/src/art/btsleepbiff.ppm
create: usr/src/art/btsleepbiff.xbm
create: usr/src/art/btsleepbiff.xpm
create: usr/src/art/btsleepmask.xbm
create: usr/src/art/btstartup2.ppm
create: usr/src/art/bttitle.ppm
create: usr/src/art/dummy.ppm
create: usr/src/art/sds.gif
create: usr/src/art/sdssmall.gif
create: usr/src/art/sdssmall.ppm
create: usr/src/audio/DevAudio.C
create: usr/src/audio/DevAudio.H
create: usr/src/audio/Makefile
create: usr/src/btref/Makefile
create: usr/src/btref/btcmds.C
create: usr/src/btref/btcmds.H
create: usr/src/btref/btcmdtab.C
create: usr/src/btref/btcmdtab.H
create: usr/src/btref/btglob.C
create: usr/src/btref/btglob.H
create: usr/src/btref/btref.C
create: usr/src/btref/btref.H
create: usr/src/btref/btsignals.H
create: usr/src/daemons/BTDBServer.C
create: usr/src/daemons/BTDBServer.H
create: usr/src/daemons/BTMDSignals.H
create: usr/src/daemons/BTMDSlave.C
create: usr/src/daemons/BTMDSlave.H
create: usr/src/daemons/BTSDClient.C
create: usr/src/daemons/BTSDClient.H
create: usr/src/daemons/BTSDSignals.H
create: usr/src/daemons/BTServer.C
create: usr/src/daemons/BTServer.H
create: usr/src/daemons/BTSlave.C
create: usr/src/daemons/BTSlave.H
create: usr/src/daemons/Makefile
create: usr/src/daemons/btserver.cf
create: usr/src/daemons/btserverd.C
create: usr/src/daemons/btserverd.H
create: usr/src/daemons/btslaved.C
create: usr/src/daemons/btslaved.H
create: usr/src/db/BTConfigFile.C
create: usr/src/db/BTConfigFile.H
create: usr/src/db/BTDB.C
create: usr/src/db/BTDB.H
create: usr/src/db/BTDBErr.C
create: usr/src/db/BTDBErr.H
create: usr/src/db/BTDBLock.C
create: usr/src/db/BTDBLock.H
create: usr/src/db/BTDBReadLock.H
create: usr/src/db/BTDBRecord.C
create: usr/src/db/BTDBRecord.H
create: usr/src/db/BTDBWriteLock.H
create: usr/src/db/BTGameStats.C
create: usr/src/db/BTGameStats.H
create: usr/src/db/BTNetwork.H
create: usr/src/db/BTNetworkEntry.C
create: usr/src/db/BTNetworkEntry.H
create: usr/src/db/BTPlayer.C
create: usr/src/db/BTPlayer.H
create: usr/src/db/BTPlayerRecord.C
create: usr/src/db/BTPlayerRecord.H
create: usr/src/db/Makefile
create: usr/src/db/ParsedFile.C
create: usr/src/db/ParsedFile.H
create: usr/src/game/.del-BTArmor.s-Oct-20-01
create: usr/src/game/BTAbout.C
create: usr/src/game/BTAbout.H
create: usr/src/game/BTArsenal.C
create: usr/src/game/BTArsenal.H
create: usr/src/game/BTBazaar.C
create: usr/src/game/BTBazaar.H
create: usr/src/game/BTBiff.C
create: usr/src/game/BTBiff.H
create: usr/src/game/BTBoard.C
create: usr/src/game/BTBoard.H
create: usr/src/game/BTBoardManager.C
create: usr/src/game/BTBoardManager.H
create: usr/src/game/BTBox.C
create: usr/src/game/BTBox.H
create: usr/src/game/BTCBoard.C
create: usr/src/game/BTCBoard.H
create: usr/src/game/BTChallenge.C
create: usr/src/game/BTChallenge.H
create: usr/src/game/BTChallengeDialog.C
create: usr/src/game/BTChallengeDialog.H
create: usr/src/game/BTCommManager.C
create: usr/src/game/BTCommManager.H
create: usr/src/game/BTComputer.C
create: usr/src/game/BTComputer.H
create: usr/src/game/BTConstants.H
create: usr/src/game/BTDebug.H
create: usr/src/game/BTDirs.H
create: usr/src/game/BTFallbacks.C
create: usr/src/game/BTGame.C
create: usr/src/game/BTGame.H
create: usr/src/game/BTLine.H
create: usr/src/game/BTList.C
create: usr/src/game/BTList.H
create: usr/src/game/BTMove.H
create: usr/src/game/BTMovePath.C
create: usr/src/game/BTMovePath.H
create: usr/src/game/BTNetManager.C
create: usr/src/game/BTNetManager.H
create: usr/src/game/BTPiece.C
create: usr/src/game/BTPiece.H
create: usr/src/game/BTPieceManager.C
create: usr/src/game/BTPieceManager.H
create: usr/src/game/BTPimp.C
create: usr/src/game/BTPimp.H
create: usr/src/game/BTProtocol.H
create: usr/src/game/BTRecon.C
create: usr/src/game/BTRecon.H
create: usr/src/game/BTRingNode.C
create: usr/src/game/BTRingNode.H
create: usr/src/game/BTRoster.C
create: usr/src/game/BTRoster.H
create: usr/src/game/BTScore.C
create: usr/src/game/BTScore.H
create: usr/src/game/BTScoreManager.C
create: usr/src/game/BTScoreManager.H
create: usr/src/game/BTSigHandlers.H
create: usr/src/game/BTSoundManager.C
create: usr/src/game/BTSoundManager.H
create: usr/src/game/BTStack.C
create: usr/src/game/BTStack.H
create: usr/src/game/BTStartup.C
create: usr/src/game/BTStartup.H
create: usr/src/game/BTStopwatch.H
create: usr/src/game/BTTimeOut.H
create: usr/src/game/BTWeapon.H
create: usr/src/game/BTWeaponManager.C
create: usr/src/game/BTWeaponManager.H
create: usr/src/game/BattleTris.C
create: usr/src/game/BattleTris.H
create: usr/src/game/Makefile
create: usr/src/game/Mem.C
create: usr/src/game/Mem.H
create: usr/src/game/PPMReader.C
create: usr/src/game/PPMReader.H
create: usr/src/game/bt_fonts.txt
create: usr/src/man/BattleTris.1
create: usr/src/man/btref.1
create: usr/src/share/BattleTris.ad
create: usr/src/share/Makefile
create: usr/src/share/btweapons.db
create: usr/src/share/btweaponsp.db
create: usr/src/signals/Makefile
create: usr/src/signals/SigHandler.H
create: usr/src/signals/SigReceiver.C
create: usr/src/signals/SigReceiver.H

usr/src/btref/btref.C
#include "BTConfig.H"


#include <iostream>
using namespace std;

#include <iomanip>
using namespace std;
#include <stdio.h>
#include <string.h>

#if HAVE_UNISTD_H
# include <unistd.h>
#endif

#if STDC_HEADERS
# include <stdlib.h>
#endif

#include "SigReceiver.H"
#include "BTDB.H"
#include "BTDBErr.H"
#include "BTDirs.H"
#include "BTProtocol.H"
#include "BTConstants.H"
#include "BTConfigFile.H"

volatile int maskint = 0;
int prompt = 0;

#include "btcmds.H"
#include "btcmdtab.H"
#include "btsignals.H"
#include "btref.H"

static const char *BTREF_CONFIGFILE = "btserver.cf";
static const char *BTREF_CMDDELIMS = " \t\n\r";
static const int BTREF_CMDBUFLEN = 255;

BTConfigFile *g_conf = 0;

BTDB *netdb = 0;
BTDB *plydb = 0;

static void parse_args(int argc, char **argv);
static void usage(char *pname);

static Command *findcommand(register char *name);
static void tokenize(char *buf, int& argc, char **argv);

void cleanup();

int main(int argc, char *argv[])
{
  parse_args(argc, argv);

  char pathbuf[1024];

  if((g_conf = new BTConfigFile(BTREF_CONFIGFILE)) == 0) {
    cerr << "btref: Insufficient memory to load" << endl;
    return BTREF_ERROR;
  }

  if(!(*g_conf)) {
    delete g_conf;
    return BTREF_ERROR;
  }

  strcpy(pathbuf, g_conf->datadir());
  strcat(pathbuf, "/");
  strcat(pathbuf, BTDB_NETWORK);

  netdb = new BTDB(pathbuf, O_CREAT | O_RDWR);

  if(!(*netdb)) {
    cerr << "btref: Failed to open network database" << endl;
    cerr << "btref: " << BTDBErrMsg(netdb->error()) << endl;
    return BTREF_ERROR;
  }

  strcpy(pathbuf, g_conf->datadir());
  strcat(pathbuf, "/");
  strcat(pathbuf, BTDB_PLAYERS);
    
  plydb = new BTDB(pathbuf, O_CREAT | O_RDWR);

  if(!(*plydb)) {
    cerr << "btref: Failed to open player database" << endl;
    cerr << "btref: " << BTDBErrMsg(netdb->error()) << endl;
    return BTREF_ERROR;
  }

  if(atexit(cleanup)) {
    cerr << "btref: Failed to install exit handler" << endl;
    return BTREF_ERROR;
  }

  SigReceiver receiver;
  receiver.reset();

  BTRefTermHandler termHdl;

  receiver.install(SIGTERM, &termHdl);
  receiver.install(SIGINT, &termHdl);
  receiver.install(SIGHUP, &termHdl);
  receiver.install(SIGQUIT, &termHdl);

  prompt = isatty(fileno(stdin));

  if(prompt)
    cout << "BattleTris Referee Software v" << BTREF_VERSION << endl;

  register Command *c;
  char cmdbuf[BTREF_CMDBUFLEN + 1];
  char *cmdargv[BTREF_CMDBUFLEN / 2];
  int cmdargc;

  for (;;) {
    if(prompt)
      cout << "btref> " << flush;

    cin.getline(cmdbuf, BTREF_CMDBUFLEN);

    if(cin.fail() || cin.eof()) {
      cout << endl;
      break;
    }

    tokenize(cmdbuf, cmdargc, cmdargv);

    if(cmdargv[0] == 0)
      continue;

    c = findcommand(cmdargv[0]);

    if(c == (Command *) -1) {
      cerr << "btref: Ambiguous command" << endl;
      continue;
    }

    if (c == 0) {
      cerr << "btref: Invalid command" << endl;
      continue;
    }

    maskint = 1;
    (*c->handler_)(cmdargc, cmdargv);
    maskint = 0;
   }
   
   return BTREF_SUCCESS;
}

static void parse_args(int argc, char **argv)
{
  extern int optind;
  extern char *optarg;
  int c;

  while(optind < argc) {
    while((c = getopt(argc, argv, "h?f:")) != (int) EOF) {
      switch(c) {

      case '?':
      case 'h':
	usage(argv[0]);
	exit(BTREF_SUCCESS);

      case 'f':
        BTREF_CONFIGFILE = optarg;
        break;

      default:
	usage(argv[0]);
	exit(BTREF_USAGE);
      }
    }

    if(optind < argc) {
      usage(argv[0]);
      exit(BTREF_USAGE);
    }
  }
}

static void usage(char *pname)
{
  cout << "Usage: " << pname << " [-h|-?] [-f configfile]\n";
  cout << "       -h, -?         Display usage information.\n";
  cout << "                      Make sure to run btref from a valid\n";
  cout << "                      BattleTris Server host.  Type \"help\"\n";
  cout << "                      at the prompt for a list of commands.\n";
  cout << "       -f configfile  Specify the pathname of the configuration\n";
  cout << "                      file which contains pathnames of files and\n";
  cout << "                      directories used by the server.\n";
  cout << flush;
}

static Command *findcommand(register char *name)
{
  register Command *c, *found = 0;
  register int nmatches = 0;
  register int longest = 0;
  register char *p, *q;

  for(c = cmdtab; p = c->name_; c++) {
    for(q = name; *q == *p++; q++) {
      if(*q == 0)
	return(c);
    }

    if(!*q) {
      if(q - name > longest) {
	longest = q - name;
	nmatches = 1;
	found = c;
      } else if(q - name == longest)
	nmatches++;
    }
  }

  if (nmatches > 1)
    return((Command *) -1);

  return(found);
}

static void tokenize(char *buf, int& argc, char **argv)
{
  register char *tok;
  argc = 0;

  for(tok = strtok(buf, BTREF_CMDDELIMS); tok != NULL;
      tok = strtok(NULL, BTREF_CMDDELIMS))
    argv[argc++] = tok;

  argv[argc] = (char *) NULL;
}

void cmd_help(int argc, char **argv)
{
  register Command *c;
  register char *arg;

  if(argc == 1) {
    cout << "== BattleTris Referee Help ================================\n";
    cout << " Commands may be abbreviated to shortest unique string\n";
    cout << " Commands which take pattern arguments use glob syntax\n";
    cout << "===========================================================\n";

    for(c = cmdtab; c < &cmdtab[CMDTABLEN]; c++)
      cout << setw(CMDMAXLEN) << c->name_ << " ~ " << c->help_ << '\n';
  } else {
    while(--argc > 0) {
      arg = *++argv;
      c = findcommand(arg);

      if(c == (Command *) -1)
	cerr << "btref: Command \"" << arg << "\" is ambiguous" << endl;
      else if (c == (Command *) 0)
	cerr << "btref: Command \"" << arg << "\" is invalid" << endl;
      else
	cout << setw(CMDMAXLEN) << c->name_ << " ~ " << c->help_ << '\n';
    }
  }

  cout << flush;
}

void cleanup()
{
  delete g_conf;
  delete netdb;
  delete plydb;
}
